File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/nodejs/test_02_task_ownership_and_failure_collection.mjs
// 任务所有权、结果复用与失败收集。
// 共同问题：异步结果能否被多个观察者复用；组合失败是否取消其他工作；
// 调用方如何选择快速失败或把失败作为结果收集。
//
// polyglot-family: async_and_concurrency
// polyglot-concept: async_await_and_result_propagation
// polyglot-related: languages/nodejs/language/test_019_async_functions_await_microtasks_and_concurrency.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('Promise.allSettled 按输入顺序收集值与失败', async () => {
  const failure = new Error('failed');

  const results = await Promise.allSettled([
    Promise.resolve(1),
    Promise.reject(failure),
  ]);

  assert.deepEqual(results, [
    { status: 'fulfilled', value: 1 },
    { status: 'rejected', reason: failure },
  ]);
});

test('Promise.all 快速拒绝但不会取消其余 Promise 工作', async () => {
  const events = [];
  let finish;
  const remaining = new Promise((resolve) => {
    finish = () => {
      events.push('remaining completed');
      resolve(2);
    };
  });

  await assert.rejects(
    Promise.all([Promise.reject(new Error('failed')), remaining]),
    /failed/,
  );
  assert.deepEqual(events, []);

  finish();
  assert.equal(await remaining, 2);
  assert.deepEqual(events, ['remaining completed']);

  // Promise 表示结果而非工作所有权；组合器的 rejection 不会自动撤销底层操作。
});

test('thenable 由 await 通过 Promise 解析协议同化', async () => {
  const events = [];
  const thenable = {
    then(resolve) {
      events.push('then');
      resolve(42);
    },
  };

  assert.equal(await thenable, 42);
  assert.deepEqual(events, ['then']);
});

File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/julia/test_02_task_ownership_and_failure_collection.jl
# polyglot-family: async_and_concurrency
# polyglot-concept: async_await_and_result_propagation
# polyglot-related: languages/julia/tooling_and_runtime/12_tasks_channels_and_cooperative_concurrency/
# polyglot-related+: test_089_task_schedule_fetch_and_states.jl
#
# 共同问题：父作用域是否拥有子任务；多个子任务失败如何被完整收集。
# 对照观察：@sync 等待词法范围内的 @async，并把失败作为 CompositeException 传播给拥有者。

using Test

@testset "@sync 建立结构化任务所有权" begin
    completed = Channel{Symbol}(2)
    @sync begin
        @async put!(completed, :first)
        @async put!(completed, :second)
    end
    @test Set([take!(completed), take!(completed)]) == Set([:first, :second])
    @test_throws CompositeException @sync @async error("owned failure")
end
